app/src/main/java/com/paranoidandroid/journey/wizard/adapters/LegsArrayAdapter.java

```diff
@@ -0,0 +1,243 @@
+    public void onBindViewHolder(final ViewHolder holder, int position) {
+        Leg leg = legs.get(position);
+
+        // Set item views based on your views and data model
+        holder.tvDestination.setText(leg.getDestination().getDisplayName());
+
+        holder.btnDeleteLeg.setOnClickListener(new View.OnClickListener() {
+            @Override
+            public void onClick(View view) {
+                onDeleteClick(holder.getAdapterPosition());
+            }
+        });
+
+        holder.btnStartDate.setOnClickListener(new View.OnClickListener() {
+            @Override
+            public void onClick(View view) {
+                onCalendarClick((Button) view, holder.getAdapterPosition(), true);
+            }
+        });
+
+        holder.btnEndDate.setOnClickListener(new View.OnClickListener() {
+            @Override
+            public void onClick(View view) {
+                onCalendarClick((Button) view, holder.getAdapterPosition(), false);
+            }
+        });
+
+        Calendar startDate = getCal(leg.getStartDate());
+
+        if (startDate == null && position > 0) {
+            Date previousEndDate = getItem(position - 1).getEndDate();
+            if (previousEndDate != null) {
+                startDate = Calendar.getInstance();
+                startDate.setTime(previousEndDate);
+                startDate.add(Calendar.DAY_OF_YEAR, 1);
+                leg.setStartDate(startDate.getTime());
+                legs.set(position, leg);
+            }
+        }
+
+        assignDate(holder.btnStartDate, startDate);
+        assignDate(holder.btnEndDate, leg.getEndDate());
+
+    }
+
+    @Override
+    public int getItemCount() {
+        return legs.size();
+    }
+
+    public void add(Leg item) {
+        legs.add(item);
+        updateParent();
+    }
+
+    public void remove(Leg item) {
+        legs.remove(item);
+        updateParent();
+    }
+
+    public void insert(Leg item, int i) {
+        legs.set(i, item);
+        updateParent();
+    }
+
+    public Leg getItem(int i) {
+        return legs.get(i);
+    }
+
+    private Context getContext() {
+        return context;
+    }
+
+    private void updateParent() {
+        notifyDataSetChanged();
+
+        List<Leg> legs = new ArrayList<>();
+        for (int i = 0; i < getItemCount(); i++){
+            legs.add(getItem(i));
+        }
+
+        Map<String, Object> result = new HashMap<>();
+        result.put(JourneyBuilder.LEGS_KEY, legs);
+
+        listener.updateJourneyData(result);
+    }
+
+    private void assignDate(Button button, Date date) {
+        assignDate(button, getCal(date));
+    }
+    /**
+     * This method either shows the selected date on the button or sets the background
+     * resource to the calendar icon.
+     */
+    private void assignDate(Button button, Calendar date) {
+        if (date == null) {
+            button.setText("");
+            AppCompatDelegate.setCompatVectorFromResourcesEnabled(true);
```

Comment:
Super minor point, but I think this only needs be called once, so could be in a static initializer.